### ISSUE TITLE
"move_append" feature

### ISSUE THREAD

**Issue by darkdragon-001** [openzfs/zfs#10446]
Concatenation of file parts is a common operation. Often, the original parts are not needed afterwards anymore.

Currently, the only possibility is to read all original part file contents, write them to a single new concatenated file and remove the parts afterwards. This is fairly inefficient.

With copy-on-write functionality in place, it should be quite easy, to create a new file descriptor and re-use the contents of the original part files. Afterwards the file descriptors of the original part files can be removed. This way, no file contents need to be copied which should speed up such operations significantly!

Therefore, I suggest introduction of a `move_append` operation.

Example:
```bash
$ ls
part_in1 part_in2 part_in3
$ zfs mv_append part_in* part_out
$ ls
part_out
```

**Comment by beren12** [comment]
I don't 't think it's that it's common at all. In the last 20 years I've maybe done it 2-3 times.

I'd be pretty interested in seeing how you do this in your patch however.

**Comment by darkdragon-001** [comment]
> I don't 't think it's that it's common at all. In the last 20 years I've maybe done it 2-3 times.

It could be used for the MinIO cloud storage provider for serving millions of files and terabytes of data every day. Users can upload large files in small chunks which are then concatenated and stored as the original file on disk (minio/minio#9657).

**Comment by beren12** [comment]
I don't think this is possible, but I look forward to seeing how you do it in your patch.

**Comment by darkdragon-001** [comment]
I have neither looked into ZFS code nor done any kernel programming at all.

For COW/snapshot/deduplication functionality, it must be possible to reference the same block of data from multiple files.

Example:
`File=Block1+Block2+Block3`; `File1=Block1`; `File2=Block2`; `File3=Block3`
This situation must be possible because I can create `File`, take a snapshot and then remove Block2+Block3 from it, resulting in `File` having the same contents as `File1`. Therefore, a way to describe this situation exists. When reversing the progress and starting with `File1`, `File2` and `File3`, it must be possible to create a file descriptor for `File` with the contents `Block1+Block2+Block3`.

**Comment by beren12** [comment]
Why do you think it must? There's not even a reflink copy in zfs. Also your example is wrong, you are changing the file so you get a new copy of it, or at least a new version of that record.

**Comment by darkdragon-001** [comment]
@beren12 The reflink copy is just not implemented but there are no fundamental problems (see https://github.com/openzfs/zfs/issues/405#issuecomment-333245135 ).

When I have a 1TB file and I change a single block, only this single block is changed and referenced in the new snapshot while the old snapshot is still pointing to the old block.

Furthermore, deduplication is supported by ZFS, so creating a new file descriptor making use of existing blocks is supported as well.

I still didn't get your point why you think it is not possible. Can you please elaborate with more details (and if possible references/links)?
